File: bloock-bridge/src/entity_mappings/anchor.rs
use bloock_core::anchor::entity::anchor::{
    Anchor as AnchorCore, AnchorNetwork as AnchorNetworkCore,
};

use crate::items::{Anchor, AnchorNetwork};

impl From<AnchorCore> for Anchor {
    fn from(a: AnchorCore) -> Self {
        Self {
            id: a.id,
            block_roots: a.block_roots,
            networks: a
                .networks
                .iter()
                .map(|network| network.clone().into())
                .collect(),
            root: a.root,
            status: a.status,
        }
    }
}

impl From<AnchorNetworkCore> for AnchorNetwork {
    fn from(n: AnchorNetworkCore) -> Self {
        Self {
            name: n.name.clone(),
            state: n.state.clone(),
            tx_hash: n.tx_hash,
        }
    }
}

impl From<AnchorNetwork> for AnchorNetworkCore {
    fn from(n: AnchorNetwork) -> Self {
        Self {
            name: n.name.clone(),
            state: n.state.clone(),
            tx_hash: n.tx_hash,
        }
    }
}

File: languages/go/client/entity/anchor.go
package entity

import "github.com/bloock/go-bridge/internal/bridge/proto"

type Anchor struct {
	Id         int64
	BlockRoots []string
	Networks   []AnchorNetwork
	Root       string
	Status     string
}

func NewAnchorFromProto(a *proto.Anchor) Anchor {
	return Anchor{
		Id:         a.Id,
		BlockRoots: a.BlockRoots,
		Networks:   MapAnchorNetworksFromProto(a.Networks),
		Root:       a.Root,
		Status:     a.Status,
	}
}

type AnchorNetwork struct {
	Name   string
	State  string
	TxHash string
}

func NewAnchorNetworkFromProto(a *proto.AnchorNetwork) AnchorNetwork {
	return AnchorNetwork{
		Name:   a.Name,
		State:  a.State,
		TxHash: a.TxHash,
	}
}

func MapAnchorNetworksFromProto(n []*proto.AnchorNetwork) []AnchorNetwork {
	networks := make([]AnchorNetwork, len(n))
	for i, network := range n {
		networks[i] = NewAnchorNetworkFromProto(network)
	}
	return networks
}

func (a AnchorNetwork) ToProto() *proto.AnchorNetwork {
	return &proto.AnchorNetwork{
		Name:   a.Name,
		State:  a.State,
		TxHash: a.TxHash,
	}
}
